design/workflow.go

```diff
@@ -0,0 +1,207 @@
+	})
+})
+
+// WorkflowOutput defines the output from a FuseML workflow
+var WorkflowOutput = Type("WorkflowOutput", func() {
+	Field(1, "name", String, "Name of the output", func() {
+		Example("prediction-url")
+	})
+	Field(2, "description", String, "Description of the output", func() {
+		Example("The URL where the exposed prediction service endpoint can be contacted to run predictions.")
+	})
+	Field(3, "type", String, "The data type of the output", func() {
+		Example("string")
+	})
+})
+
+// WorkflowStep defines a step for a FuseML workflow
+var WorkflowStep = Type("WorkflowStep", func() {
+	Field(1, "name", String, "The name of the step", func() {
+		Example("predictor")
+	})
+	Field(2, "image", String, "The image used to execute the step", func() {
+		Example("ghcr.io/fuseml/kfserving-predictor:1.0")
+	})
+	Field(3, "inputs", ArrayOf(WorkflowStepInput), "List of inputs for the step")
+	Field(4, "outputs", ArrayOf(WorkflowStepOutput), "List of output from the step")
+	Field(5, "env", ArrayOf(StepEnv), "List of environment variables available for the container running the step")
+})
+
+// WorkflowStepInput defines the input for a FuseML workflow step
+var WorkflowStepInput = Type("WorkflowStepInput", func() {
+	Field(1, "name", String, "Name of the input", func() {
+		Example("model-uri")
+	})
+	Field(2, "value", String, "Value of the input", func() {
+		Example("s3://mlflow-artifacts/3/c7ae3b0e6fd44b4b96f7066c66672551/artifacts/model")
+	})
+	Field(3, "codeset", StepInputCodeset, "Codeset associated with the input")
+})
+
+// StepInputCodeset defines the Codeset type of input for a FuseML workflow step
+var StepInputCodeset = Type("StepInputCodeset", func() {
+	Field(1, "name", String, "Name or ID of the codeset", func() {
+		Example("mlflow-project")
+	})
+	Field(2, "path", String, "Path where the codeset will be mounted inside the container running the step", func() {
+		Example("/project")
+	})
+})
+
+// WorkflowStepOutput defines the output from a FuseML workflow step
+var WorkflowStepOutput = Type("WorkflowStepOutput", func() {
+	Field(1, "name", String, "Name of the variable to hold the step output value", func() {
+		Example("model-uri")
+	})
+	Field(2, "image", StepOutputImage, "If the step builds a container image as output it will be referenced as 'image'")
+})
+
+// StepOutputImage defines the output from a FuseML workflow when it builds a container image
+var StepOutputImage = Type("StepOutputImage", func() {
+	Field(1, "dockerfile", String, "Path to the Dockerfile used to build the image", func() {
+		Example("/project/.fuseml/Dockerfile")
+	})
+	Field(2, "name", String, "Name of the image, including the repository where the image will be stored", func() {
+		Example("registry.fuseml-registry/mlflow-project/mlflow-codeset:0.1")
+	})
+})
+
+// StepEnv defines the environment variables that are loaded inside the container running a FuseML workflow step
+var StepEnv = Type("StepEnv", func() {
```

Comment:
This could be modeled as a `MapOf(String, String)` in the parent object. Something we can take care of when we align the definition of workflows to that of runnables.